### PR TITLE
[MT-59]:(fix) Improve photos pending items feedback

### PR DESCRIPTION
**Pull request by PixoDev**
Improved pending photos feedback.

- Added retry on fail mechanism for uploaded photos
- Recover missing previews (for locally deleted photos mostly)

### CODE REVIEW

File: src/contexts/Photos/sync.ts [hunk 1]
```diff
@@ -23,8 +26,24 @@ export const startSync = async (handlers: SyncHandlers) => {
   }
 
   const syncManager = photos.sync;
+  let pendingPhotosInterval: number | null = null;
+  syncManager.onPhotosItemUploadStart(handlers.onPhotosItemUploadStart);
+  syncManager.onPhotosItemUploadProgress(handlers.onPhotosItemUploadProgress);
 
+  const startPendingPhotosUpdatePolling = () => {
```
Comment (PixoDev) [inline src/contexts/Photos/sync.ts:33]: With this, we provide initial feedback quickly, every 500ms, after the first Photos item is uploaded, this interval is deactivated, and the pending items are updated each time a Photos item is uploaded

File: src/contexts/Photos/Photos.context.tsx [hunk 2]
```diff
@@ -314,14 +350,15 @@ export const PhotosContextProvider: React.FC = ({ children }) => {
...
   }
 
-  function handleRemovePhotosItems(photosItems: PhotosItem[]) {
-    syncedPhotosItems.current = syncedPhotosItems.current.filter((mergedPhotosItem) =>
-      photosItems.find((pi) => pi.name === mergedPhotosItem.name) ? false : true,
-    );
+  async function handleRemovePhotosItems(photosItems: PhotosItem[]) {
+    for (const photosItem of photosItems) {
+      if (photosItem.photoId) {
+        await photos.database.deleteSyncedPhotosItem(photosItem.photoId);
```
Comment (sg-gs) [inline src/contexts/Photos/Photos.context.tsx:356]: Maybe a bulk delete could be a good improvement on speed/resources than this, just using:

```
const photosWithId = photosItems.filter((photo) => !!photo.photoId);
for (let i = 0; i < photosWithId.length; i += 20) {
    await photos.database.deleteSyncedPhotosItems(nPhotos.slice(i, i+20));
}
```

File: src/services/photos/network/PhotosNetworkManager.ts [hunk 3]
```diff
@@ -31,33 +36,60 @@ export class PhotosNetworkManager implements RunnableService<PhotosNetworkManage
...
-    this.sdk = sdk;
+  constructor(options = { enableLog: ENABLE_PHOTOS_NETWORK_MANAGER_LOGS }) {
     this.options = options;
-    this.queue.drain(() => {
+  }
+
+  private createQueue() {
+    const queue = async.queue<PhotosNetworkOperation, Photo | null, Error>((task, next) => {
+      if (this.isAborted) {
+        throw new AbortedOperationError();
```
Comment (sg-gs) [inline src/services/photos/network/PhotosNetworkManager.ts:53]: Is better to follow the same flow that the queue establishes, with is according to the async [docs](https://caolan.github.io/async/v3/docs.html#QueueObject) using the next callback with an error, in this case:

```
// maybe adding queue.kill() too could be a good idea.
return next(new AbortedOperationError());
```

File: src/services/photos/network/PhotosNetworkManager.ts [hunk 4]
```diff
@@ -193,7 +256,7 @@ export class PhotosNetworkManager implements RunnableService<PhotosNetworkManage
   }
 
   getPendingTasks() {
-    return this.queue.length();
+    return this.queue.length() + this.queue.running();
```
Comment (sg-gs) [inline src/services/photos/network/PhotosNetworkManager.ts:259]: nice one

File: src/services/photos/sync/devicePhotosScanner/devicePhotosScanner.ts [hunk 5]
```diff
@@ -76,13 +73,12 @@ export class DevicePhotosScannerService extends RunnableService<DevicePhotosScan
     this.updateStatus(DevicePhotosScannerStatus.PAUSED);
   }
 
-  public isDone() {
+  public hasFinished() {
```
Comment (sg-gs) [inline src/services/photos/sync/devicePhotosScanner/devicePhotosScanner.ts:76]: Good naming improvement!

File: src/types/photos.ts [hunk 6]
```diff
@@ -146,6 +146,7 @@ export enum PhotosNetworkOperationResult {
 export interface PhotosNetworkOperation {
   photosItem: PhotosItem;
   lastError?: Error;
+  retrys: number;
```
Comment (sg-gs) [inline src/types/photos.ts:149]: typo here: _retries_

**Review by sg-gs — APPROVED** [review]

File: src/contexts/Photos/Photos.context.tsx (hunk 2 above)
Comment (PixoDev) [reply src/contexts/Photos/Photos.context.tsx:356]: yep, should be done using a transaction, need to go deeper into that

**Pull request merged**